packages/playwright-test/src/worker/workerMain.ts

```diff
@@ -386,100 +382,96 @@ export class WorkerMain extends ProcessRunner {
+        testInfo._timeoutManager.setCurrentRunnable({ type: 'test', slot: afterHooksSlot });
         debugTest(`tearing down test scope started`);
         const testScopeError = await testInfo._runFn(() => this._fixtureRunner.teardownScope('test', testInfo._timeoutManager));
         debugTest(`tearing down test scope finished`);
         firstAfterHooksError = firstAfterHooksError || testScopeError;
-        // Attribute to 'teardown' because worker fixtures are not perceived as a part of a test.
-        testInfo._timeoutManager.setCurrentRunnable({ type: 'teardown', slot: teardownSlot });
-        debugTest(`tearing down worker scope started`);
-        const workerScopeError = await testInfo._runFn(() => this._fixtureRunner.teardownScope('worker', testInfo._timeoutManager));
-        debugTest(`tearing down worker scope finished`);
-        firstAfterHooksError = firstAfterHooksError || workerScopeError;
       });
-    }
 
-    afterHooksStep.complete({ error: firstAfterHooksError });
+      if (testInfo._isFailure())
+        this._isStopped = true;
+
+      if (this._isStopped) {
+        // Run all remaining "afterAll" hooks and teardown all fixtures when worker is shutting down.
+        // Mark as "cleaned up" early to avoid running cleanup twice.
+        this._didRunFullCleanup = true;
+
+        // Give it more time for the full cleanup.
+        await testInfo._runWithTimeout(async () => {
+          debugTest(`running full cleanup after the failure`);
+          for (const suite of reversedSuites) {
+            const afterAllError = await this._runAfterAllHooksForSuite(suite, testInfo);
+            firstAfterHooksError = firstAfterHooksError || afterAllError;
+          }
+          const teardownSlot = { timeout: this._project.timeout, elapsed: 0 };
+          // Attribute to 'test' so that users understand they should probably increate the test timeout to fix this issue.
+          testInfo._timeoutManager.setCurrentRunnable({ type: 'test', slot: teardownSlot });
+          debugTest(`tearing down test scope started`);
+          const testScopeError = await testInfo._runFn(() => this._fixtureRunner.teardownScope('test', testInfo._timeoutManager));
+          debugTest(`tearing down test scope finished`);
+          firstAfterHooksError = firstAfterHooksError || testScopeError;
+          // Attribute to 'teardown' because worker fixtures are not perceived as a part of a test.
+          testInfo._timeoutManager.setCurrentRunnable({ type: 'teardown', slot: teardownSlot });
+          debugTest(`tearing down worker scope started`);
+          const workerScopeError = await testInfo._runFn(() => this._fixtureRunner.teardownScope('worker', testInfo._timeoutManager));
+          debugTest(`tearing down worker scope finished`);
+          firstAfterHooksError = firstAfterHooksError || workerScopeError;
+        });
+      }
+      if (firstAfterHooksError)
+        step.complete({ error: firstAfterHooksError });
```

Comment:
testInfo._runAsStep will call step.complete({}) again, is that expected?

Reply:
Yes, that is alright.